pkg/reporterreporter/report.go

```diff
@@ -30,49 +41,143 @@ type RequestData struct {
 	Data      reporter_v1.NamespaceReport `json:"data"`
 }
 
-func (r *Report) Run(ctx context.Context) error {
-	namespaces, err := r.KubernetesClient.CoreV1().Namespaces().List(meta_v1.ListOptions{})
+func NewReport(slurperURI string, cluster string, reporterClient client.Interface, kubeClient kubernetes.Interface, logger *zap.Logger) ctrl.Server {
+	logger.Info("ReporterReporter configured",
+		zap.String("cluster_id", cluster),
+		zap.String("slurper_uri", slurperURI),
+		zap.String("slurper_path", slurperPath))
+
+	return &report{
+		clusterID:        cluster,
+		httpClient:       util.HTTPClient(),
+		kubernetesClient: kubeClient,
+		logger:           logger,
+		mutator: restclient.NewRequestMutator(
+			restclient.BaseURL(slurperURI),
+			restclient.JoinPath(slurperPath),
+			restclient.Method(http.MethodPost)),
+		reporterClient: reporterClient,
+	}
+}
+
+func isRetriableError(statusCode int) bool {
+	_, ok := map[int]bool{
+		http.StatusRequestTimeout:  true,
+		http.StatusTooManyRequests: true,
+	}[statusCode]
+
+	if ok {
+		return true
+	}
+
+	return statusCode >= 500 && statusCode <= 599
+}
+
+// sendData will send the requestData containing the report to slurper
+func (r *report) sendData(requestData RequestData) (retriable bool, err error) {
```

Comment:
Why not pass context down to this level and include a reference in the rest client?

Reply:
I've added it; I didn't add it because it wasn't really being used, but I guess it's good for the future.